src/mr_allocator.rs

```diff
@@ -1,60 +1,722 @@
+/// Alloc extent memory with `ibv_reg_mr()`
+static RDMA_ALLOC_EXTENT_HOOK: unsafe extern "C" fn(
+    extent_hooks: *mut extent_hooks_t,
+    new_addr: *mut c_void,
+    size: usize,
+    alignment: usize,
+    zero: *mut i32,
+    commit: *mut i32,
+    arena_ind: u32,
+) -> *mut c_void = extent_alloc_hook;
+
+/// Custom extent dalloc hook used by jemalloc
+/// Dalloc extent memory with `ibv_dereg_mr()`
+static RDMA_DALLOC_EXTENT_HOOK: unsafe extern "C" fn(
+    extent_hooks: *mut extent_hooks_t,
+    addr: *mut c_void,
+    size: usize,
+    committed: i32,
+    arena_ind: u32,
+) -> i32 = extent_dalloc_hook;
+
+/// Custom extent merge hook used by jemalloc
+/// Merge two adjacent extents to a bigger one
+static RDMA_MERGE_EXTENT_HOOK: unsafe extern "C" fn(
+    extent_hooks: *mut extent_hooks_t,
+    addr_a: *mut c_void,
+    size_a: usize,
+    addr_b: *mut c_void,
+    size_b: usize,
+    committed: i32,
+    arena_ind: u32,
+) -> i32 = extent_merge_hook;
+
+/// Custom extent hooks
+static mut RDMA_EXTENT_HOOKS: extent_hooks_t = extent_hooks_t {
+    alloc: Some(RDMA_ALLOC_EXTENT_HOOK),
+    dalloc: Some(RDMA_DALLOC_EXTENT_HOOK),
+    destroy: None,
+    commit: None,
+    decommit: None,
+    purge_lazy: None,
+    purge_forced: None,
+    split: None,
+    merge: Some(RDMA_MERGE_EXTENT_HOOK),
 };
 
+lazy_static! {
+    /// Default extent hooks of jemalloc
+    static ref ORIGIN_HOOKS: extent_hooks_t = get_default_hooks_impl(DEFAULT_ARENA_INDEX);
+    /// The correspondence between extent metadata and `raw_mr`
+    #[derive(Debug)]
+    pub(crate) static ref EXTENT_TOKEN_MAP: Arc<ParkingLotMutex<BTreeMap<usize, Item>>> = Arc::new(ParkingLotMutex::new(BTreeMap::<usize, Item>::new()));
+    /// The correspondence between `arena_ind` and `ProtectionDomain`
+    pub(crate) static ref ARENA_PD_MAP: Arc<LockFreeCuckooHash<u32, Arc<ProtectionDomain>>> = Arc::new(LockFreeCuckooHash::new());
+}
+
+/// Combination between extent metadata and `raw_mr`
+#[derive(Debug)]
+pub(crate) struct Item {
+    /// addr of an extent(memory region)
+    addr: usize,
+    /// length of an extent(memory region)
+    len: usize,
+    /// Reference of RawMemoryRegion
+    raw_mr: Arc<RawMemoryRegion>,
+    /// arena index of this extent
+    arena_ind: u32,
+}
+
 /// Memory region allocator
 #[derive(Debug)]
 pub(crate) struct MrAllocator {
     /// Protection domain that holds the allocator
-    pd: Arc<ProtectionDomain>,
+    _pd: Arc<ProtectionDomain>,
+    /// Arena index
+    arena_ind: u32,
 }
 
 impl MrAllocator {
     /// Create a new MR allocator
     pub(crate) fn new(pd: Arc<ProtectionDomain>) -> Self {
-        Self { pd }
+        #[allow(clippy::expect_used)]
+        let arena_ind =
+            init_je_statics(Arc::<ProtectionDomain>::clone(&pd)).expect("init je statics failed");
+        Self { _pd: pd, arena_ind }
     }
 
     /// Allocate a MR according to the `layout`
+    #[allow(clippy::as_conversions)]
```

Comment:
Please use "cast", not "as"

Reply:
`*mut u8` cannot be converted to `usize` by `cast()`